rust-runtime/aws-smithy-http-server/src/runtime_error.rs

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Runtime error type.
+//!
+//! This module contains [`RuntimeError`] type.
+//!
+//! As opposed to rejection types (see [`crate::rejection`]), which are an internal detail about
+//! the framework, `RuntimeError` is surfaced to clients in HTTP responses: indeed, it implements
+//! [`axum_core::response::IntoResponse`]. Rejections can be "grouped" and converted into a
+//! specific `RuntimeError` kind: for example, all request rejections due to serialization issues
+//! can be conflated under the [`RuntimeErrorKind::Serialization`] enum variant.
+//!
+//! The HTTP response representation of the specific `RuntimeError` can be protocol-specific: for
+//! example, the runtime error in the RestJson1 protocol sets the `X-Amzn-Errortype` header.
+//!
+//! Generated code works always works with [`crate::rejection`] types when deserializing requests
+//! and serializing response. Just before a response needs to be sent, the generated code looks up
+//! and converts into the corresponding `RuntimeError`, and then it uses the its
+//! [`axum_core::response::IntoResponse`] implementation to render and send a response.
+
+use crate::protocols::Protocol;
+
+#[derive(Debug)]
+pub enum RuntimeErrorKind {
+    // UnknownOperation,
+    /// Request failed to deserialize or response failed to serialize.
+    Serialization(crate::Error),
```

Comment:
What about calling this Deserialization? It can be confusing otherwise.

Reply:
This error is used both when a request fails to deserialize and when a response fails to serialize; AWS protocols use `SerializationException` for both.